pkg/expression/builtin_cast.go

```diff
@@ -2525,7 +2525,7 @@ func BuildCastFunctionWithCheck(ctx BuildContext, expr Expression, tp *types.Fie
 	// since we may reset the flag of the field type of CastAsJson later which
 	// would affect the evaluation of it.
 	if tp.EvalType() != types.ETJson && err == nil {
-		res = FoldConstant(ctx, res)
+		return FoldConstant(ctx, res), nil
```

Comment:
FoldConstant will copy the res. but here overwrite the ```res```. it will lead the problem.